apm-lambda-extension/main_test.go

```diff
@@ -0,0 +1,343 @@
+	InvokeMultipleTransactionsOverload MockEventType = "MultipleTransactionsOverload"
+	Shutdown                           MockEventType = "Shutdown"
+)
+
+type APMServerBehavior string
+
+const (
+	TimelyResponse APMServerBehavior = "TimelyResponse"
+	SlowResponse   APMServerBehavior = "SlowResponse"
+	Hangs          APMServerBehavior = "Hangs"
+	Crashes        APMServerBehavior = "Crashes"
+)
+
+type MockEvent struct {
+	Type              MockEventType
+	APMServerBehavior APMServerBehavior
+	ExecutionDuration float64
+	Timeout           float64
+}
+
+func processMockEvent(currId string, event MockEvent, APMServer *httptest.Server) {
+	sendLogEvent(currId, "platform.start")
+	client := http.Client{}
+	switch event.Type {
+	case InvokeHang:
+		time.Sleep(time.Duration(event.Timeout) * time.Second)
+	case InvokeStandard:
+		time.Sleep(time.Duration(event.ExecutionDuration) * time.Second)
+		req, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+		res, _ := client.Do(req)
+		log.Println(res.StatusCode)
+	case InvokeStandardFlush:
+		time.Sleep(time.Duration(event.ExecutionDuration) * time.Second)
+		reqData, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events?flushed=true", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+		client.Do(reqData)
+	case InvokeWaitgroupsRace:
+		time.Sleep(time.Duration(event.ExecutionDuration) * time.Second)
+		reqData0, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+		reqData1, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+		go client.Do(reqData0)
+		go client.Do(reqData1)
+		time.Sleep(650 * time.Microsecond)
+	case InvokeMultipleTransactionsOverload:
+		wg := sync.WaitGroup{}
+		for i := 0; i < 200; i++ {
+			go func() {
+				wg.Add(1)
+				time.Sleep(time.Duration(event.ExecutionDuration) * time.Second)
+				reqData, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+				client.Do(reqData)
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+	case Shutdown:
+		reqData, _ := http.NewRequest("POST", "http://localhost:8200/intake/v2/events", bytes.NewBuffer([]byte(event.APMServerBehavior)))
+		client.Do(reqData)
+	}
+	sendLogEvent(currId, "platform.runtimeDone")
+}
+
+func sendNextEventInfo(id string, event MockEvent) []byte {
+	nextEventInfo := extension.NextEventResponse{
+		EventType:          "INVOKE",
+		DeadlineMs:         time.Now().UnixNano()/int64(time.Millisecond) + int64(event.Timeout*1000),
+		RequestID:          id,
+		InvokedFunctionArn: "arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test",
+		Tracing:            extension.Tracing{},
+	}
+	if event.Type == Shutdown {
+		nextEventInfo.EventType = "SHUTDOWN"
+	}
+
+	out, _ := json.Marshal(nextEventInfo)
```

Comment:
can we use [`json.Encoder`](https://pkg.go.dev/encoding/json#Encoder) here? stuart was suggesting we use `json.Decoder` instead of `Unmarshal` directly.

Reply:
Solved in [d9a8edc](https://github.com/elastic/apm-aws-lambda/pull/136/commits/d9a8edc0ca29cd924a0217ed22f9715a5c270c90)